stdlib/public/core/UnicodeScalarProperties.swift

```diff
@@ -0,0 +1,1449 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+// Exposes advanced properties of Unicode.Scalar defined by the Unicode
+// Standard.
+//===----------------------------------------------------------------------===//
+
+import SwiftShims
+
+extension Unicode.Scalar {
+
+  /// A value that provides access to properties of a Unicode scalar that are
+  /// defined by the Unicode standard.
+  public struct Properties {
+    internal init(_scalar: Unicode.Scalar) {
+      // We convert the value to the underlying UChar32 type here and store it
+      // in that form to make calling the ICU APIs cleaner below.
+      self._value = __swift_stdlib_UChar32(bitPattern: _scalar._value)
+
+      // Likewise, we cache the UTF-16 encoding of the scalar for a few of the
+      // properties that require treating it like a string (e.g., case
+      // mappings).
+      let utf16 = _scalar.utf16
+      self._utf16Length = utf16.count
+      self._utf16 = (utf16[0], _utf16Length > 1 ? utf16[1] : 0)
+    }
+
+    internal var _value: __swift_stdlib_UChar32
+    internal var _utf16: (UTF16.CodeUnit, UTF16.CodeUnit)
+    internal var _utf16Length: Int
```

Comment:
Encoding a scalar into UTF-16 is pretty trivial and almost certainly not worth storing an extra word or two to avoid re-computing it. It's also not worth doing every time a user writes `myScalar.properties.isWhitespace`. I suggest computing this lazily.

Reply:
Sounds good. Since you said above that you're working on a fix for the case mappings, I'll hold off on fixing this until that's ready.

Reply:
Do you think you could cherry-pick this: https://github.com/milseman/swift/commit/0fdd3d79dc33bc55798de392af2aa096c59563c9 ?

I tried to open a PR against this PR, but I couldn't find your repo in the list...

Reply:
Do you have any tests? I want to make sure I didn't break anything

Reply:
Thanks! These new case mappings look like they did the trick (I've pushed an update with your changes, and also removed the UTF-16 caching).

I don't have any proper unit tests, I've just been spot-testing with the REPL. I'll get those together next.